### PR TITLE
[Merged by Bors] - feat(AlgebraicGeometry/ResidueField): classification of `Spec K ⟶ X` with `K` a field

### PR DESCRIPTION
**Pull request by chrisflav**
This PR adds more API for residue fields and in particular the classification of morphisms `Spec K ⟶ X` with `K` a field.

From the valuative criterion project.

Co-authored-by: Qi Ge
Co-authored-by: Andrew Yang

---

- [x] depends on: #15240 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/AlgebraicGeometry/ResidueField.lean [hunk 1]
```diff
@@ -165,23 +191,102 @@ lemma residue_residueFieldCongr (X : Scheme) {x y : X} (h : x = y) :
...
+  infer_instance
 
 @[reassoc (attr := simp)]
 lemma residueFieldCongr_fromSpecResidueField {x y : X} (h : x = y) :
     Spec.map (X.residueFieldCongr h).hom ≫ X.fromSpecResidueField _ =
       X.fromSpecResidueField _ := by
   subst h; simp
 
+@[reassoc]
+lemma Hom.residueFieldMap_fromSpecResidueField (x : X) :
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ResidueField.lean:219]:
```suggestion
lemma Hom.Spec_map_residueFieldMap_fromSpecResidueField (x : X) :
```
Or is this too long?

File: Mathlib/AlgebraicGeometry/Morphisms/Preimmersion.lean [hunk 2]
```diff
@@ -95,6 +96,25 @@ theorem comp_iff {X Y Z : Scheme} (f : X ⟶ Y) (g : Y ⟶ Z) [IsPreimmersion g]
...
+    (h₁ : Embedding (PrimeSpectrum.comap f)) (h₂ : f.SurjectiveOnStalks) :
+    IsPreimmersion (Spec.map f) where
+  base_embedding := h₁
+  surj_on_stalks (x : PrimeSpectrum S) := by
+    let e := Scheme.arrowStalkMapSpecIso f x
+    haveI : (RingHom.toMorphismProperty <| fun f ↦ Function.Surjective f).RespectsIso := by
+      rw [← RingHom.toMorphismProperty_respectsIso_iff]
+      exact RingHom.surjective_respectsIso
+    apply ((RingHom.toMorphismProperty <| fun f ↦ Function.Surjective f).arrow_mk_iso_iff e).mpr
+    exact h₂ x.asIdeal x.isPrime
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/Preimmersion.lean:109]: Is it hard to make this into an iff?

**Review by erdOne — COMMENTED** [review]
And maybe wait for #17403 before merging this

File: Mathlib/AlgebraicGeometry/Morphisms/Preimmersion.lean (hunk 2 above)
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/Preimmersion.lean]: No, done.

**Comment by chrisflav** [comment]
> And maybe wait for #17403 before merging this

I agree. In the meantime, I have untangled the imports because of the `large-import` label and also because I think it is reasonable that `ResidueField` does not import `ClosedImmersion` but rather the other way round.

**Comment by erdOne** [comment]
Please merge master and incorporate the new changes. Thanks!

File: Mathlib/AlgebraicGeometry/ResidueField.lean [hunk 3]
```diff
@@ -20,12 +20,14 @@ The following are in the `AlgebraicGeometry.Scheme` namespace:
 - `AlgebraicGeometry.Scheme.Hom.residueFieldMap`: A morphism of schemes induce a homomorphism of
   residue fields.
 - `AlgebraicGeometry.Scheme.fromSpecResidueField`: The canonical map `Spec κ(x) ⟶ X`.
+- `AlgebraicGeometry.Scheme.SpecToEquivOfField`: morphisms `Spec K ⟶ X` for a field `K` correspond
```
Comment (jcommelin) [inline Mathlib/AlgebraicGeometry/ResidueField.lean:23]: This moved to a new file, right?

File: Mathlib/AlgebraicGeometry/Morphisms/ClosedImmersion.lean [hunk 4]
```diff
@@ -133,6 +131,15 @@ theorem of_comp {X Y Z : Scheme} (f : X ⟶ Y) (g : Y ⟶ Z) [IsClosedImmersion
...
 
+instance {X : Scheme.{u}} (x) : IsClosedImmersion (Spec.map (X.residue x)) :=
+  IsClosedImmersion.spec_of_surjective (X.residue x)
+    Ideal.Quotient.mk_surjective
+
+instance {X : Scheme.{u}} (x : X) : IsPreimmersion (X.fromSpecResidueField x) := by
+  dsimp only [Scheme.fromSpecResidueField]
+  rw [IsPreimmersion.comp_iff]
+  infer_instance
+
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/Morphisms/ClosedImmersion.lean:142]: I think this can go into `ResidueField` along with `SpecToEquivOfField`.

Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/ClosedImmersion.lean:143]: Not sure, this depends on the `IsClosedImmersion` instance of `Spec.map (X.residue x)` and we can't import `IsClosedImmersion` in `ResidueField`. The alternative would be to add `IsPreimmersion (Spec.map (X.residue x))` in `ResidueField` separately and only later upgrade it to `IsClosedImmersion`, but not sure if this is really better than the current version.

Comment (erdOne) [reply Mathlib/AlgebraicGeometry/Morphisms/ClosedImmersion.lean:143]: I feel that the latter makes more sense to me. As this result itself has little to do with closed immersions.

File: Mathlib/AlgebraicGeometry/ResidueField.lean (hunk 3 above)
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/ResidueField.lean:23]: No longer after implementing Andrews suggestion.

File: Mathlib/AlgebraicGeometry/Morphisms/ClosedImmersion.lean (hunk 4 above)
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/Morphisms/ClosedImmersion.lean]: Okay, I did that and added a comment on the `IsPreimmersion` instance to find the stronger result.

File: Mathlib/AlgebraicGeometry/ResidueField.lean [hunk 5]
```diff
@@ -171,23 +199,101 @@ lemma residue_residueFieldCongr (X : Scheme) {x y : X} (h : x = y) :
...
+
+lemma descResidueField_fromSpecResidueField {K : Type*} [Field K] (X : Scheme) {x}
+    (f : X.presheaf.stalk x ⟶ .of K) [IsLocalHom f] :
+    Spec.map (X.descResidueField f) ≫
+      X.fromSpecResidueField x = Spec.map f ≫ X.fromSpecStalk x := by
+  simp [fromSpecResidueField, ← Spec.map_comp_assoc]
+
+instance {R : CommRingCat} [LocalRing R] (f : Spec R ⟶ X) :
+    IsLocalHom (F := X.presheaf.stalk _ →+* R) (stalkClosedPointTo f) :=
+  isLocalHom_stalkClosedPointTo f
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ResidueField.lean:258]: Is this still necessary?

File: Mathlib/AlgebraicGeometry/ResidueField.lean [hunk 6]
```diff
@@ -171,23 +199,101 @@ lemma residue_residueFieldCongr (X : Scheme) {x y : X} (h : x = y) :
...
+  rw [Category.assoc, ← Spec_map_stalkMap_fromSpecStalk, ← Spec.map_comp_assoc,
+    ← Spec.map_comp_assoc]
+  rfl
+
+@[simp]
+lemma fromSpecResidueField_apply (x : X.carrier) (s : Spec (X.residueField x)) :
+    (X.fromSpecResidueField x).base s = x := by
+  simp [fromSpecResidueField]
+
+lemma range_fromSpecResidueField  (x : X.carrier) :
```
Comment (erdOne) [inline Mathlib/AlgebraicGeometry/ResidueField.lean:240]:
```suggestion
lemma range_fromSpecResidueField (x : X.carrier) :
```

**Comment by erdOne** [comment]
LGTM other than these minor comments
But I think I wrote most of the statements so I'll refrain from maintainer-merging.

File: Mathlib/AlgebraicGeometry/ResidueField.lean (hunk 5 above)
Comment (chrisflav) [reply Mathlib/AlgebraicGeometry/ResidueField.lean]: Indeed, it is not.

**Comment by joelriou** [comment]
Thanks!

bors merge

**Pull request closed without merging**